### ISSUE TITLE
Different rendering between input tag help and Razor helper for nullable bool

### ISSUE THREAD

**Issue by skimedic** [aspnet/Mvc#5667]
I searched, but I didn't see anything about this.  If it's already been logged and/or handled, I apologize in advance.  The `<input asp-for="NullableBool"></input>` tag helper renders:

```html
<input name="NullableBool" id="NullableBool" type="text" value="">
```

and the `Html.InputFor(x=>x.NullableBool)` razor helper renders:

```html
<select name="NullableBool" class="list-box tri-state" id="NullableBool"><option selected="selected" value="">Not Set</option><option value="true">True</option><option value="false">False</option></select>
```

Is this by design? It's a completely different user experience.

As a side note, thought not as drastic a change, the non-nullable bool values are rendered differently as well:
Non-nullable with razor helper:

```html
<input name="PostiveBool" class="check-box" id="PostiveBool" type="checkbox" checked="checked" value="true" data-val-required="The PostiveBool field is required." data-val="true"><input name="PostiveBool" type="hidden" value="false">
```

Non-nullable with Core MVC Tag Helper:

```html
<input name="PostiveBool" id="PostiveBool" type="checkbox" checked="checked" value="true"><input name="PostiveBool" type="hidden" value="false">
```

Test Code:
Model:

```c#
    public class ModelToTestTemplates
    {
        public bool? NullableBool { get; set; }
        public bool PostiveBool { get; set; }
        public bool NegativeBool { get; set; }
    }
```

View:
```html
@model SpyStore.MVC.Models.ModelToTestTemplates
@Html.EditorFor(model => model.NullableBool)
@Html.EditorFor(model => model.PostiveBool)
@Html.EditorFor(model => model.NegativeBool)
<hr/>
<input asp-for="NullableBool"/><br/>
<input asp-for="PostiveBool"/><br/>
<input asp-for="NegativeBool"/><br/>
```

Action method:
```c#
        public ActionResult Test()
        {
            return View(new ModelToTestTemplates {NegativeBool = false, PostiveBool = true});
        }
```

Please let me know if there is anything else you need from me.  tia.

**Comment by dougbu** [comment]
Bottom line is HTML and tag helpers cover some overlapping features but are not identical. In particular, no tag helpers support the extensibility (templates) or recursion features of the `Html.Display[For]()` and `Html.Editor[For]()` HTML helpers.

> Is this by design? It's a completely different user experience.

Yes, this is by design. The `<input>` tag helper _never_ results in a different element (though it can generate an additional `<input>` as shown in your later sample). On the other hand, the `Html.Display[For]()` and `Html.Editor[For]()` HTML helpers can generate `<input>` or `<select>` elements and more.

> the non-nullable bool values are rendered differently as well

This is by design though the biggest change your example shows is due to the HTML helper for the `PostiveBool` property coming before the tag helper for that property. MVC generates client-side validation attributes once per `<form>`.

The other difference (not adding `class="check-box"`) is also by design. The default templates for `Html.Display[For]()` and `Html.Editor[For]()` assume Bootstrap.

**Comment by dougbu** [comment]
Please reopen if I missed something you asked or you believe there's an issue in here somewhere.

**Comment by skimedic** [comment]
Thank you for taking the time to explain all of that.  I guess where I am hung up (and not sure how to document in my book) is where the bool? renders a text box when using tag helpers and not a check box.  Won't that wreck havoc on model binding?  If the user enters "yes", or "foo", or anything other than 1/0 or true/false, in the text box instead of checking a check box, then model binding will fail, but I'm not sure that's really the fault of the user since we gave them a text box.  The solution is to use an editor template, then it all goes away.  But I did find it interested that bool? renders a text box input  control.  Again, if that's by design, then I'll just stick with using display and editor templates to control the rendering.
